crates/store/re_chunk_store/src/writes.rs

```diff
@@ -150,29 +204,145 @@ impl ChunkStore {
             );
         }
 
-        let event = ChunkStoreEvent {
-            store_id: self.id.clone(),
-            store_generation: self.generation(),
-            event_id: self
-                .event_id
-                .fetch_add(1, std::sync::atomic::Ordering::Relaxed),
-            diff: ChunkStoreDiff::addition(Arc::clone(chunk)),
+        let events = if self.config.enable_changelog {
+            let events: Vec<_> = diffs
+                .into_iter()
+                .map(|diff| ChunkStoreEvent {
+                    store_id: self.id.clone(),
+                    store_generation: self.generation(),
+                    event_id: self
+                        .event_id
+                        .fetch_add(1, std::sync::atomic::Ordering::Relaxed),
+                    diff,
+                })
+                .collect();
+
+            Self::on_events(&events);
+
+            events
+        } else {
+            Vec::new()
         };
 
-        {
-            let events = &[event.clone()];
+        Ok(events)
+    }
 
-            if cfg!(debug_assertions) {
-                let any_event_other_than_addition = events
-                    .iter()
-                    .any(|e| e.kind != ChunkStoreDiffKind::Addition);
-                assert!(!any_event_other_than_addition);
-            }
+    /// Finds the most appropriate candidate for compaction.
```

Comment:
This is interesting though the heuristic nature makes it hard to reason through how it works in practice.

Given chunks will almost arrive in order given `log_time` / `log_seq` that seems like it will always create a bias toward 2 votes for the trivial arrival-order compaction. My suspicion is that we would be better off bias toward compacting along the user-defined timeline when there is one, rather than the arrival order since that is likely to be the natural view and most likely timeline for range queries.

An additional observation is that chunk-overlap seems like it should be one of the main drivers of compaction. Any time 2 chunks overlap on one or more timelines, it will cause performance issues for us.

If we have an opportunity to prevent an overlap from being created through compaction, that seems like it will always be a net win.

Reply:
Yeah there's definitely an infinite stream of possible improvements in this space.

I'm hoping this simple vote system can help us quickly experiment with more complex biases as we go ("you overlap? +5 to you!", "you're a user-defined timeline? +4 to you!").